api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java

```diff
@@ -539,6 +589,29 @@ protected void handleParticipantReference(Criteria criteria, ReferenceAndListPar
 		}
 	}
 	
+	protected Optional<Criterion> findMatchingConcepts(@NotNull String conceptAlias, TokenAndListParam concepts,
+	        Criteria criteria) {
+		if (concepts == null) {
+			return Optional.empty();
+		}
+		
+		return handleAndListParamBySystem(concepts, (system, tokens) -> {
+			if (system.isEmpty()) {
+				return Optional.of(or(
+				    in(String.format("%s.conceptId", conceptAlias),
+				        tokensToParams(tokens).map(NumberUtils::toInt).collect(Collectors.toList())),
+				    in(String.format("%s.uuid", conceptAlias), tokensToList(tokens))));
+			} else {
+				if (!containsAlias(criteria, "cm")) {
+					criteria.createAlias(String.format("%s.conceptMappings", conceptAlias), "cm")
+					        .createAlias("cm.conceptReferenceTerm", "crt");
+				}
+				
+				return Optional.of(generateSystemQuery(system, tokensToList(tokens)));
+			}
+		});
```

Comment:
Can we refactor out the lambda in here (the expression beginning `(system, tokens) ->`) and turn it into a proper method which can be called both here and `handleResourceCode`

Reply:
Also refactor `handleResourceCode` to be called `handleCodeableConcept`. Hopefully we can eventually get rid of the `TokenOrListParam` version of that.

Reply:
@ibacher So I guess we'll need to change the parameter either here or in handleResourceCode to make it `TokenOrListParam` or `TokenAndListParam`

Reply:
Well, I'd like to keep them separate for now so that anything that's relying on `TokenOrListParam` doesn't break and we keep the scope of this PR relatively focused. We can have two method with the same name that are distinguished by parameter types, so that shouldn't be an issue. We just need to make these two resemble each other so it's clear that they provide the same functionality.